drivers/isl/isl.cpp

```diff
@@ -0,0 +1,350 @@
+  write_reg(0x22, 0x0A); //Fixed Error
+  write_reg(0x24, 0x48); //Cross Talk I MSB
+  write_reg(0x25, 0x54); //Cross Talk I LSB
+  write_reg(0x26, 0x36); //Cross Talk I EXP
+  write_reg(0x27, 0x45); //Cross Talk Q MSB
+  write_reg(0x28, 0xB0); //Cross Talk Q LSB
+  write_reg(0x29, 0xB1); //Cross Talk Q EXP
+  write_reg(0x2A, 0xFC); //Cross Talk Gain MSB
+  write_reg(0x2B, 0xCA); //Cross Talk Gain LSB
+  write_reg(0x2C, 0x07); //Mag EXP
+  write_reg(0x2D, 0x9B); //Mag MSB
+  write_reg(0x2E, 0xD0); //Mag LSB
+  write_reg(0x2F, 0x12); //Phase Offset MSB
+  write_reg(0x30, 0x15); //Phase Offset LSB
+  write_reg(0x60, 0x01); //Interrupt Control
+  write_reg(0x61, 0x00); //Data Invalid Msk
+  write_reg(0x62, 0x00); //Detection Control
+  write_reg(0x63, 0x00); //Detection Cnd 1
+  write_reg(0x64, 0x00); //Detection Cnd 2
+  write_reg(0x70, 0x00); //Dist Th 1 Hgh MSB
+  write_reg(0x71, 0xDB); //Dist Th 1 Hgh LSB
+  write_reg(0x72, 0xFF); //Dist Th 1 Low MSB
+  write_reg(0x73, 0x9E); //Dist Th 1 Low LSB
+  write_reg(0x74, 0x05); //Dist Th 2 Hgh MSB
+  write_reg(0x75, 0x61); //Dist Th 2 Hgh LSB
+  write_reg(0x76, 0x00); //Dist Th 2 Low MSB
+  write_reg(0x77, 0xB1); //Dist Th 2 Low LSB
+  write_reg(0x78, 0x0F); //Dist Th 3 Hgh MSB
+  write_reg(0x79, 0x5E); //Dist Th 3 Hgh LSB
+  write_reg(0x7A, 0x03); //Dist Th 3 Low MSB
+  write_reg(0x7B, 0xD7); //Dist Th 3 Low LSB
+  write_reg(0x7D, 0x95); //Mag Th 1 Hgh
+  write_reg(0x7E, 0x02); //Mag Th 1 EXP
+  write_reg(0x7F, 0xC7); //Mag Th 1 Low
+  write_reg(0x81, 0xA3); //Mag Th 2 Hgh
+  write_reg(0x82, 0x06); //Mag Th 2 EXP
+  write_reg(0x83, 0x94); //Mag Th 2 Low
+  write_reg(0x84, 0x02); //Motion Dist Th MSB
+  write_reg(0x85, 0x4E); //Motion Dist Th LSB
+  write_reg(0x86, 0x00); //Motion MAG Th
+  write_reg(0x90, 0x0F); //Driver Range
+  write_reg(0x91, 0x80); //Emitter DAC
+  write_reg(0x92, 0x00); //Driver Control
+  write_reg(0x93, 0x00); //Thresh DAC
+  write_reg(0x94, 0x00); //Driver Boost
+  write_reg(0x96, 0x00); //Driv Chg Bal DAC
+  write_reg(0x97, 0x00); //Front End Ctl
+  write_reg(0x98, 0x00); //AFE Ctl Regs
+  write_reg(0xA5, 0x00); //Emitter ADC Offset
+  write_reg(0xA7, 0x00); //Temp Sens Reg A
+  write_reg(0xA8, 0x00); //Temp Sens Reg B
+  write_reg(0xA9, 0x00); //Temp Sens Reg C
+  write_reg(0xAA, 0x00); //Temp Sens ADC Mode
+  write_reg(0xae, 0x00); //Internal Reset
+  write_reg(0x31, 0x00); //Temp Reference
+  write_reg(0x33, 0x00); //Phase Exponent
+  write_reg(0x34, 0x00); //Temp Coeff 1
+  write_reg(0x39, 0x00); //Temp Coeff 2
+  write_reg(0x36, 0x00); //Ambient Coeff 1
+  write_reg(0x3B, 0x00); //Ambient Coeff 2
+  unsigned char sample_delay_reg = read_reg(0x10); // Sample Len
+  _isl_sample_delay = calc_sample_delay(sample_delay_reg);
+  return 0;
+}
+
+//Calculate sample period as per value received from Sensor
+//calculations based on information provided by intersil
+int ISL::calc_sample_delay(unsigned char value)
+{
+  float d = 0.0711 * pow(2, value);
```

Comment:
This operation is problematic in two ways: You're using double precision, which is a software emulated mode. And you're using the pow() call for squaring a number, which is extremely expensive.
Last but not least you're passing the argument as unsigned char, which will require *MORE* operations than using using a float or int, because the processor (depending on the resulting byte code) needs to mask the unused bits and properly truncate the smaller type.

The right way to do this is like this:
```C
int ISL::calc_sample_delay(float value)
{
int d = 0.0711f * value * value;
return d + 1;
}
```
Depending on what your intent was with the integer (how you intended to use the floor operation (or not)) this could be done in a single line.